### PR TITLE
Refactor parser: Avoid some assertion for StmtNode

### PR DESCRIPTION
**Pull request by shenli**
Make code cleaner. No performance improvement.

Before:
> ➜  parser git:(master) ✗ go test -run=XXX -bench=. -benchtime=10s
BenchmarkSysbenchSelect-4   	 3000000	      4589 ns/op	    1888 B/op	      27 allocs/op
BenchmarkParse-4            	 1000000	     14310 ns/op	    5224 B/op	      89 allocs/op
PASS
ok  	github.com/pingcap/tidb/parser	32.910s
➜  parser git:(master) ✗ go test -run=XXX -bench=. -benchtime=10s
BenchmarkSysbenchSelect-4   	 3000000	      4770 ns/op	    1888 B/op	      27 allocs/op
BenchmarkParse-4            	 1000000	     14001 ns/op	    5224 B/op	      89 allocs/op
PASS
ok  	github.com/pingcap/tidb/parser	33.489s
➜  parser git:(master) ✗ go test -run=XXX -bench=. -benchtime=10s
BenchmarkSysbenchSelect-4   	 3000000	      4679 ns/op	    1888 B/op	      27 allocs/op
BenchmarkParse-4            	^[[A 1000000	     14028 ns/op	    5224 B/op	      89 allocs/op
PASS
ok  	github.com/pingcap/tidb/parser	32.791s
➜  parser git:(master) ✗ go test -run=XXX -bench=. -benchtime=10s
BenchmarkSysbenchSelect-4   	 3000000	      4628 ns/op	    1888 B/op	      27 allocs/op
BenchmarkParse-4            	 1000000	     14434 ns/op	    5224 B/op	      89 allocs/op
PASS
ok  	github.com/pingcap/tidb/parser	33.395s

After:

> ➜  parser git:(shenli/parser) ✗ go test -run=XXX -bench=. -benchtime=10s
> BenchmarkSysbenchSelect-4   	 3000000	      4586 ns/op	    1888 B/op	      27 allocs/op
> BenchmarkParse-4            	 1000000	     14229 ns/op	    5224 B/op	      89 allocs/op
> PASS
> ok  	github.com/pingcap/tidb/parser	32.819s
> ➜  parser git:(shenli/parser) ✗ go test -run=XXX -bench=. -benchtime=10s
> BenchmarkSysbenchSelect-4   	 3000000	      4567 ns/op	    1888 B/op	      27 allocs/op
> BenchmarkParse-4            	 1000000	     14253 ns/op	    5224 B/op	      89 allocs/op
> PASS
> ok  	github.com/pingcap/tidb/parser	32.768s
> ➜  parser git:(shenli/parser) ✗ go test -run=XXX -bench=. -benchtime=10s
> BenchmarkSysbenchSelect-4   	 3000000	      4622 ns/op	    1888 B/op	      27 allocs/op
> BenchmarkParse-4            	 1000000	     14979 ns/op	    5224 B/op	      89 allocs/op
> PASS
> ok  	github.com/pingcap/tidb/parser	33.719s
> ➜  parser git:(shenli/parser) ✗ go test -run=XXX -bench=. -benchtime=10s
> BenchmarkSysbenchSelect-4   	 3000000	      4701 ns/op	    1888 B/op	      27 allocs/op
> BenchmarkParse-4            	 1000000	     14687 ns/op	    5224 B/op	      89 allocs/op
> PASS
> ok  	github.com/pingcap/tidb/parser	33.645s
> ➜  parser git:(shenli/parser) ✗ go test -run=XXX -bench=. -benchtime=10s
> BenchmarkSysbenchSelect-4   	 3000000	      4561 ns/op	    1888 B/op	      27 allocs/op
> BenchmarkParse-4            	 1000000	     14233 ns/op	    5224 B/op	      89 allocs/op
> PASS
> ok  	github.com/pingcap/tidb/parser	32.730s

### CODE REVIEW

**Comment by zz-jason** [comment]
LGTM

File: parser/parser.y [hunk 1]
```diff
@@ -412,21 +413,61 @@ import (
...
+	SelectStmt			"SELECT statement"
+	RenameTableStmt         	"rename table statement"
+	ReplaceIntoStmt			"REPLACE INTO statement"
+	RevokeStmt			"Revoke statement"
+	RollbackStmt			"ROLLBACK statement"
+	SetStmt				"Set variable statement"
+	ShowStmt			"Show engines/databases/tables/columns/warnings/status statement"
+	Statement			"statement"
+	ExplainableStmt			"explainable statement"
+	TruncateTableStmt	"TRANSACTION TABLE statement"
```
Comment (XuHuaiyu) [inline parser/parser.y]: align these 5 lines.

File: parser/parser.y [hunk 2]
```diff
@@ -4804,15 +4810,15 @@ Statement:
...
 |	ShowStmt
-|	TruncateTableStmt
-|	UpdateStmt
-|	UseStmt
 |	SubSelect
 	{
 		// `(select 1)`; is a valid select statement
 		// TODO: This is used to fix issue #320. There may be a better solution.
-		$$ = $1.(*ast.SubqueryExpr).Query
+		$$ = $1.(*ast.SubqueryExpr).Query.(ast.StmtNode)
```
Comment (XuHuaiyu) [inline parser/parser.y]: why we add type assertion here?

Comment (shenli) [reply parser/parser.y]: $1.(*ast.SubqueryExpr).Query is a ResultSet, not StmtNode.

**Comment by shenli** [comment]
@XuHuaiyu PTAL

**Comment by XuHuaiyu** [comment]
circleci failed @shenli

**Comment by shenli** [comment]
@XuHuaiyu I can not open CircleCI now. I merge master and rerun it.

**Pull request merged**